lib/babe/babe.go

```diff
@@ -467,6 +467,30 @@ func (b *Service) handleSlot(epoch, slotNum uint64,
 		return errNilParentHeader
 	}
 
+	atGenesisBlock := b.blockState.GenesisHash().Equal(parentHeader.Hash())
+	if !atGenesisBlock {
+		bestBlockSlotNum, err := b.blockState.GetSlotForBlock(parentHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("could not get slot for block %s: %w", parentHeader.Hash(), err)
+		}
+
+		if bestBlockSlotNum > slotNum {
+			return errLaggingSlot
+		}
+
+		if bestBlockSlotNum == slotNum {
+			// pick parent of best block instead to handle slot
+			newParentHeader, err := b.blockState.GetHeader(parentHeader.ParentHash)
+			if err != nil {
+				return fmt.Errorf("could not get header for hash %s: %w", parentHeader.ParentHash, err)
+			}
+			if newParentHeader == nil {
+				return errNilParentHeader
+			}
+			parentHeader = newParentHeader
```

Comment:
in the previous line when I am returning the error, I am using hash from previous parentHeader, 
If I use it directly, in case of error, parentHeader will become nil and I won't be able to get its parentHeader.parentHash. Plus it also looks cleaner.